core/apple/src/BuffersApple.kt

```diff
@@ -0,0 +1,67 @@
+@file:OptIn(UnsafeNumber::class)
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import kotlinx.io.bytestring.ByteString
+import kotlinx.io.bytestring.buildByteString
+import platform.Foundation.*
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerMax
+import platform.posix.memcpy
+import platform.posix.uint8_tVar
+
+internal fun Buffer.write(source: CPointer<uint8_tVar>?, maxLength: Int) {
+    var currentOffset = 0
+    while (currentOffset < maxLength) {
+        val tail = writableSegment(1)
+
+        val toCopy = minOf(maxLength - currentOffset, Segment.SIZE - tail.limit)
+        tail.data.usePinned {
+            memcpy(it.addressOf(tail.pos), source + currentOffset, toCopy.convert())
+        }
+
+        currentOffset += toCopy
+        tail.limit += toCopy
+    }
+    size += maxLength
+}
+
+internal fun Buffer.readAtMostTo(sink: CPointer<uint8_tVar>?, maxLength: Int): Int {
+    val s = head ?: return 0
+    val toCopy = minOf(maxLength, s.limit - s.pos)
+    s.data.usePinned {
+        memcpy(sink, it.addressOf(s.pos), toCopy.convert())
+    }
+
+    s.pos += toCopy
+    size -= toCopy.toLong()
+
+    if (s.pos == s.limit) {
+        head = s.pop()
+        SegmentPool.recycle(s)
+    }
+
+    return toCopy
+}
+
+internal fun Buffer.snapshotAsNSData(): NSData {
+    if (size == 0L) return NSData.data()
+
+    check(size.toULong() <= NSUIntegerMax) { "Buffer is too long ($size) to be converted into NSData." }
+
+    val data = NSMutableData.create(length = size.convert())!!
```

Comment:
Does it make sense to allocate a block of memory, fill it and then return wrap into `NSData` (https://developer.apple.com/documentation/foundation/nsdata/1416020-init) instead of creating and returning `NSMutableData`?
Or can one always cast a pointer to `NSData` to `NSMutableData` and modify the underlying data anyway? I'm sorry for my ignorance about that.

Reply:
You can't necessarily cast any `NSData` as `NSMutableData` to modify it. You could get a cast exception. So yes, initializing `NSData` with pre-allocated bytes would prevent the cast. Although since this is a snapshot copy of the buffer state, there's not really an issue if the caller did decide to modify the bytes, since it's a copy they own. It's also always possible to modify the underlying bytes of `NSData` with lower-level APIs, even if it is supposed to be immutable.

But another advantage of allocating the bytes with `malloc` would be that it doesn't bother zeroing them out as [`NSMutableData.dataWithLength:`](https://developer.apple.com/documentation/foundation/nsmutabledata/1547233-datawithlength#return_value) does, so this could have a slight performance advantage. I'll go ahead and make that change.